### PR TITLE
Item Components for HoverEvents (fixes #3688)

### LINKED ISSUE #3688: Sending components of an item in HoverEvent is not possible
### Bungeecord version

BungeeCord-Bootstrap:1.20-R0.3-SNAPSHOT:52ab21b:1844

### Server version

_No response_

### Client version

1.20.6

### Bungeecord plugins

n/a

### The bug

When creating a HoverEvent with a Item content the components of that item are not shown on the client (this also applies to spigot) I tried this on spigot with
```
            String itemTag = stack.getItemMeta().getAsString();
            msg.setHoverEvent(new HoverEvent(Action.SHOW_ITEM, new Item(stack.getType().getKey().toString(), 1, itemTag != null ? ItemTag.ofNbt(itemTag) : null)));
```
(i also tried stack.getItemMeta().getAsComponentString()) - both don't work, only the item type is shown but no enchantments, custom name or whatever. It was working in 1.20.4.

### Log output (links)

_No response_

### Checking

- [X] I am using BungeeCord and **not a fork**. Issues with forks should not be reported here.
- [X] I think this is **not** an issue with a bungeecord plugin.
- [ ] I have not read these checkboxes and therefore I just ticked them all.
- [X] This is not a question or plugin creation help request.

### PR DESCRIPTION
**Pull request by OstlerDev**
## Item HoverEvents for 1.20.5+
I went through and implemented a simple way to use the new Item components system for HoverEvent's until a better solution that does not use `serializeItemAsJson` can be implemented ([and that also doesn't use nms...](https://github.com/snowgears/Shop/commit/9940d7775155e4153c62aea543c0532c532c1f2a)) by serializing the components some other way. I felt like this was a good compromise for now that both adds the functionality and makes it simple to use and expose.

Luckily, it was a minor change because of the robustness of the existing Content serialization system, so great job there, I was quite surprised at how easy it actually was to add.

Let me know if you would like to see any changes, I am more than happy to change anything you like as I get time.

## Example Usage
```java
JsonObject itemJson = Bukkit.getUnsafe().serializeItemAsJson(ItemStack item);
// {"id":"minecraft:elytra","count":1,"components":{"minecraft:enchantments":{"minecraft:mending":1},"minecraft:repair_cost":1},"DataVersion":4440}
new HoverEvent(HoverEvent.Action.SHOW_ITEM, new Item(itemJson));
// supports an easy one-liner
new HoverEvent(HoverEvent.Action.SHOW_ITEM, new Item(Bukkit.getUnsafe().serializeItemAsJson(item)));

// alternatively use a custom built json string to initialize
JsonObject jsonItem = JsonParser.parseString("""
{
    "id": "minecraft:elytra",
    "count": 1,
    "components": {
        "minecraft:enchantments": {
            "minecraft:mending": 1
        },
        "minecraft:repair_cost": 5
    }
}
""");
new HoverEvent(HoverEvent.Action.SHOW_ITEM, new Item(jsonItem));
```

Screenshot 2025-08-12 at 6 00 58 PM
Screenshot 2025-08-12 at 6 02 42 PM

## Copilot Summary
This pull request updates the item hover content system to support Minecraft's new Item Components format (introduced in 1.20.5+) alongside the legacy NBT tag format. It refactors the `Item` class to add constructors and fields for components, updates serialization/deserialization logic to correctly handle both formats, and adds comprehensive test coverage to verify compatibility and correctness for both legacy and modern item formats.

**Item hover content system modernization:**

* Added support for the new Item Components system in the `Item` class, including a new `components` field, multiple constructors for modern and legacy formats, and updated default values for `count` and `id`. Legacy NBT tag usage is now deprecated. (`chat/src/main/java/net/md_5/bungee/api/chat/hover/content/Item.java`) [[1]](diffhunk://#diff-b3bb2649a4b8655b179e16182a18403ff9c2030a3aab3c0449a7e998e8960c10L3-L11) [[2]](diffhunk://#diff-b3bb2649a4b8655b179e16182a18403ff9c2030a3aab3c0449a7e998e8960c10L24-R131)

**Serialization and deserialization improvements:**

* Refactored `ItemSerializer` to correctly serialize and deserialize items using either the legacy NBT tag format or the new components format. This includes handling different key names for `count` and default values for `id` depending on the format. (`serializer/src/main/java/net/md_5/bungee/api/chat/hover/content/ItemSerializer.java`) [[1]](diffhunk://#diff-778737a051061a8f1bfd859299053f1e2c7fc576f09399daffa888101bdf9f84R16-L26) [[2]](diffhunk://#diff-778737a051061a8f1bfd859299053f1e2c7fc576f09399daffa888101bdf9f84L50-R96)

**Expanded test coverage:**

* Added comprehensive tests to verify correct serialization, deserialization, and round-trip conversion for both legacy and modern item formats, including edge cases for null values and complex nested components. (`serializer/src/test/java/net/md_5/bungee/api/chat/ComponentsTest.java`) [[1]](diffhunk://#diff-a3608d564a777bf3fd8c293fa42abfb7a62b715aee247232133542df29907373R5-R14) [[2]](diffhunk://#diff-a3608d564a777bf3fd8c293fa42abfb7a62b715aee247232133542df29907373R103-R269)

### CODE REVIEW

**Comment by md-5** [comment]
Thanks, seems worthwhile to me but let's see if any others have comments on the approach

**Comment by Outfluencer** [comment]
Shouldn't the components field be de/serialized in the ItemSerializer?

**Comment by OstlerDev** [comment]
> Shouldn't the components field be de/serialized in the ItemSerializer?

@Outfluencer I just added the components field to ItemSerializer and also added some tests. You are right that we needed to add it to the serializer, thank you for pointing that out!

(p.s. also cleaned up the style and verified it locally, so hopefully the pipelines should succeed this time)

File: serializer/src/main/java/net/md_5/bungee/api/chat/hover/content/ItemSerializer.java [hunk 1]
```diff
@@ -49,6 +49,7 @@ public Item deserialize(JsonElement element, Type type, JsonDeserializationConte
         return new Item(
                 ( value.has( "id" ) ) ? value.get( "id" ).getAsString() : null,
                 count,
+                ( value.has( "components" ) ) ? context.deserialize( value.get( "components" ), JsonElement.class ) : null,
```
Comment (Outfluencer) [inline serializer/src/main/java/net/md_5/bungee/api/chat/hover/content/ItemSerializer.java:52]: is deserialize required here as it already is a JsonElement, i tested it locally and all checks passed without it

File: serializer/src/main/java/net/md_5/bungee/api/chat/hover/content/ItemSerializer.java [hunk 2]
```diff
@@ -62,6 +63,10 @@ public JsonElement serialize(Item content, Type type, JsonSerializationContext c
         {
             object.addProperty( "Count", content.getCount() );
         }
+        if ( content.getComponents() != null )
+        {
+            object.add( "components", context.serialize( content.getComponents() ) );
```
Comment (Outfluencer) [inline serializer/src/main/java/net/md_5/bungee/api/chat/hover/content/ItemSerializer.java:68]: same here with serialize

**Comment by OstlerDev** [comment]
One thing I did notice was that the default for `id` is no longer allowed to be `minecraft:air` and it will throw an error if it is air. I would propose we default to `minecraft:stone` to maintain backwards compatibility where HoverEvents were allowed to display ItemHovers with a null `id` field.

Additionally, it looks like the count/amount field changed from being `Count` to being `count`, my new change will also address that.

I have a commit I am about to push up with extra tests and documentation.

**Comment by OstlerDev** [comment]
@Outfluencer I believe it should be complete and ready now. Maybe I will add an additional test to verify that the null values for id properly get set to their defaults, but I feel quite confident as is currently with the amount of test coverage.

(also: I saw I accidentally put `dirt` in the serializer, I am switching that to `stone`, just fyi since you might notice it)

**Comment by OstlerDev** [comment]
Just fixed the `id` default to be `stone` for modern, and added a couple more tests to get 100% coverage on all changes and give me just a little more confidence haha. Note: I can't seem to get my ssh key signing properly working for commits right now, my apologies.

Edit: On a different note, I will see if I can add something to Bukkit/CraftBukkit/Spigot so that ItemStack properly exposes a function to either get the correct components json string or actual components json object. Obviously deeper changes like that take more time, but it should be able to be fully compatible with the system we are implementing for components right now I believe.

File: chat/src/main/java/net/md_5/bungee/api/chat/hover/content/Item.java [hunk 3]
```diff
@@ -23,10 +23,101 @@ public class Item extends Content
...
+     * ItemTag tag = ItemTag.ofNbt(item.getItemMeta().getAsString());
+     * new HoverEvent(HoverEvent.Action.SHOW_ITEM, new Item(item.getType().getKey().toString(), item.getAmount(), tag));
+     * </pre>
+     * @param id The item id. Defaults to: 'minecraft:air'
+     * @param count The item count.
+     * @param tag The item tag.
+     * @deprecated Since 1.20.5+ you need to use the new constructors for the Item Components system instead {@link Item#Item(JsonObject)} or {@link Item#Item(String, int, JsonElement)}
+     */
+    @Deprecated
+    public Item(String id, int count, ItemTag tag)
```
[the remark this answers is not in the document]
Comment (OstlerDev) [reply chat/src/main/java/net/md_5/bungee/api/chat/hover/content/Item.java:48]: Maybe it is as simple as switching the other constructor to `public Item(String id, JsonElement components, int count)` which would prevent mismatches, but I dislike changing the order of the arguments... Maybe a better Java wizard than I knows what to do there.

[the remark this answers is not in the document]
Comment (OstlerDev) [reply chat/src/main/java/net/md_5/bungee/api/chat/hover/content/Item.java]: I changed it over to `public Item(String id, JsonElement components, int count)` for now to prevent overlap of the constructors

**Comment by OstlerDev** [comment]
I noticed [in the new documentation that the default should be 1 for count](https://minecraft.wiki/w/Data_component_format#Usage) so I went ahead and updated the serializer to match along with a test to cover that functionality :)

I also solved an issue with passing `null` for the legacy ItemTag arg since I am pretty sure I saw people doing that previously. I opened a conversation about that above.

**Comment by OstlerDev** [comment]
I believe everything should be complete, I double checked it and everything seems good to me.

Any thoughts by anybody?

**Comment by Outfluencer** [comment]
Currently not working as passing a json object seems like breaking the current tag utils as there are LazilyParsedNumbers that we don't handle currently

For example this: 

```
ComponentBuilder builder = new ComponentBuilder( "HALLO" );
builder.event(new HoverEvent(HoverEvent.Action.SHOW_ITEM, new Item(JsonParser.parseString("{" +
        "      \"id\": \"minecraft:elytra\"," +
        "      \"count\": 1," +
        "      \"components\": {" +
        "          \"minecraft:custom_name\": \"Sky Wings\"," +
        "          \"minecraft:enchantments\": {" +
        "              \"minecraft:mending\": 99" +
        "          }" +
        "      }" +
        "  }").getAsJsonObject())));
con.sendMessage(builder.build());
```

Results in an encoder exception

```
Caused by: java.lang.IllegalArgumentException: Unknown JSON primitive: 99
	at net.md_5.bungee.protocol.util.TagUtil.fromJson(TagUtil.java:157)
	at net.md_5.bungee.protocol.util.TagUtil.fromJson(TagUtil.java:163)
	at net.md_5.bungee.protocol.util.TagUtil.fromJson(TagUtil.java:163)
	at net.md_5.bungee.protocol.util.TagUtil.fromJson(TagUtil.java:163)
	at net.md_5.bungee.protocol.util.TagUtil.fromJson(TagUtil.java:163)
	at net.md_5.bungee.protocol.util.TagUtil.fromJson(TagUtil.java:231)
	at net.md_5.bungee.protocol.util.TagUtil.fromJson(TagUtil.java:163)
	at net.md_5.bungee.protocol.DefinedPacket.writeBaseComponent(DefinedPacket.java:192)
	at net.md_5.bungee.protocol.packet.SystemChat.write(SystemChat.java:34)
	at net.md_5.bungee.protocol.DefinedPacket.write(DefinedPacket.java:610)
	at net.md_5.bungee.protocol.MinecraftEncoder.encode(MinecraftEncoder.java:27)
	at net.md_5.bungee.protocol.MinecraftEncoder.encode(MinecraftEncoder.java:10)
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:107)
	... 44 more	
```

**Comment by Outfluencer** [comment]
Could fix that issue by adding 

```
 if (number instanceof LazilyParsedNumber) {
    return new DoubleTag(number.doubleValue());
}
```


to TagUtil#fromJson(JsonElement)

**Comment by md-5** [comment]
Why is that example giving a LazilyParsedNumber? I thought that was only meant to happen with string numbers, ie "1" but I don't see that in the example

**Comment by Outfluencer** [comment]
I don't know
seems like 
```
JsonParser.parseString("{" +
        "      \"id\": \"minecraft:elytra\"," +
        "      \"count\": 1," +
        "      \"components\": {" +
        "          \"minecraft:custom_name\": \"Sky Wings\"," +
        "          \"minecraft:enchantments\": {" +
        "              \"minecraft:mending\": 99" +
        "          }" +
        "      }" +
        "  }").getAsJsonObject()
```
 just results in having it in it for me even without 99 being a string

**Comment by md-5** [comment]
I think we need to be a bit smarter rather than always making it a double.
Eg,

```diff
--- a/protocol/src/main/java/net/md_5/bungee/protocol/util/TagUtil.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/util/TagUtil.java
@@ -5,6 +5,9 @@ import com.google.gson.JsonElement;
 import com.google.gson.JsonNull;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonPrimitive;
+import com.google.gson.internal.LazilyParsedNumber;
+import com.google.gson.internal.NumberLimits;
+import java.math.BigDecimal;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.LinkedHashMap;
@@ -32,6 +35,27 @@ import net.md_5.bungee.nbt.type.StringTag;
 public final class TagUtil
 {
 
+    private static Number lazyToBoxed(LazilyParsedNumber lazy)
+    {
+        BigDecimal decimal = NumberLimits.parseBigDecimal( lazy.toString() );
+
+        if ( decimal.signum() == 0 || decimal.scale() <= 0 || decimal.stripTrailingZeros().scale() <= 0 )
+        {
+            long longVal = decimal.longValueExact();
+
+            if ( (int) longVal == longVal )
+            {
+                return (int) longVal;
+            } else
+            {
+                return longVal;
+            }
+        } else
+        {
+            return decimal.doubleValue();
+        }
+    }
+
     private static byte nbtTypeOfJson(JsonElement json)
     {
         if ( json instanceof JsonPrimitive )
@@ -40,6 +64,10 @@ public final class TagUtil
             if ( jsonPrimitive.isNumber() )
             {
                 Number number = json.getAsNumber();
+                if ( number instanceof LazilyParsedNumber )
+                {
+                    number = lazyToBoxed( (LazilyParsedNumber) number );
+                }
 
                 if ( number instanceof Byte )
                 {
@@ -121,6 +149,10 @@ public final class TagUtil
             if ( jsonPrimitive.isNumber() )
             {
                 Number number = json.getAsNumber();
+                if ( number instanceof LazilyParsedNumber )
+                {
+                    number = lazyToBoxed( (LazilyParsedNumber) number );
+                }
 
                 if ( number instanceof Byte )
                 {
```

**Comment by Janmm14** [comment]
It is the default strategy of gson to use LazilyParsedNumbers.

It was just never as problem, because bungeecord does have its own (de-)serializers and uses functions like `getAsInt()` and did not yet attempt to forward anything to TagUtil that was just a string parsed as json - it would always deserialize json, and serialize from our data classes, thats why it was not a problem yet, e.g. with the `count`.

Did anyone tried to send a chat message from the server through bungeecord to the client with such an enchantment? From my understanding, our `ItemTag` implementation might be problematic already related to `LazilyParsedNumber`s due to its he stringified json field `nbt`.

I guess we can not simply treat all these numbers as doubles, becuase I suspect minecraft might not like getting a double for the enchantment level...

So if my speculations are right, we need to change ItemTag to contain a JsonValue to keep rough type information for unknown numeric values.
After that, we could check the toString() of the lazily parsed number to guess whether we should treat it as int or double.

Sorry for only writing theory here, pls some1 else test.

Or maybe changing the default gson number-to-number-strategy might already help, not sure about potential type problems in the nbt-conversion tho.

**Comment by md-5** [comment]
> It is the default strategy of gson to use LazilyParsedNumbers.
> 
> It was just never as problem, because bungeecord does have its own (de-)serializers and uses functions like `getAsInt()` and did not yet attempt to forward anything to TagUtil that was just a string parsed as json - it would always deserialize json, and serialize from our data classes, thats why it was not a problem yet, e.g. with the `count`.
> 
> Did anyone tried to send a chat message from the server through bungeecord to the client with such an enchantment? From my understanding, our `ItemTag` implementation might be problematic already related to `LazilyParsedNumber`s due to its he stringified json field `nbt`.
> 
> I guess we can not simply treat all these numbers as doubles, becuase I suspect minecraft might not like getting a double for the enchantment level...
> 
> So if my speculations are right, we need to change ItemTag to contain a JsonValue to keep rough type information for unknown numeric values. After that, we could check the toString() of the lazily parsed number to guess whether we should treat it as int or double.
> 
> Sorry for only writing theory here, pls some1 else test.
> 
> Or maybe changing the default gson number-to-number-strategy might already help, not sure about potential type problems in the nbt-conversion tho.

I think my snippet above should address at least some of those issues
